packages/plugin-legacy/src/index.ts

```diff
@@ -331,24 +331,30 @@ function viteLegacyPlugin(options: Options = {}): Plugin[] {
         }
 
         return (chunkInfo) => {
-          let fileName =
+          const file =
             typeof fileNames === 'function' ? fileNames(chunkInfo) : fileNames
+          const fileArr = file.split('/')
+          let fileName = fileArr.pop()!
 
           if (fileName.includes('[name]')) {
             // [name]-[hash].[format] -> [name]-legacy-[hash].[format]
             fileName = fileName.replace('[name]', '[name]-legacy')
-          } else if (fileName.includes('[hash]')) {
-            // custom[hash].[format] -> [name]-legacy[hash].[format]
-            // custom-[hash].[format] -> [name]-legacy-[hash].[format]
-            // custom.[hash].[format] -> [name]-legacy.[hash].[format]
-            fileName = fileName.replace(/[.-]?\[hash\]/, '-legacy$&')
           } else {
-            // entry.js -> entry-legacy.js
-            // entry.min.js -> entry-legacy.min.js
-            fileName = fileName.replace(/(.+?)\.(.+)/, '$1-legacy.$2')
+            if (fileName.includes('[hash]') && !fileName.startsWith('[hash]')) {
+              // custom[hash].[format] -> [name]-legacy[hash].[format]
+              // custom-[hash].[format] -> [name]-legacy-[hash].[format]
+              // custom.[hash].[format] -> [name]-legacy.[hash].[format]
+              fileName = fileName.replace(/[.-]?\[hash\]/, '-legacy$&')
```

Comment:
I think we should support other separators here (using `[^\w\d]` instead of `[.-]`), but I'll send another PR to do that, so we keep this one strictly about improving leading `[hash]` in the pattern.